### ISSUE TITLE
New rule: UnnecessaryFullyQualifiedName

### ISSUE THREAD

**Issue by 3flex** [arturbosch/detekt#1501]
## Expected Behavior of the rule
Identify unnecessary fully qualified names. E.g. `kotlin.io.println()` can be simplified to `println()`.

Suggest starting with the default imports provided by Kotlin, i.e.:
* kotlin.*
* kotlin.annotation.*
* kotlin.collections.*
* kotlin.comparisons.*
* kotlin.io.*
* kotlin.ranges.*
* kotlin.sequences.*
* kotlin.text.*
* java.lang.*
* kotlin.jvm.*
* kotlin.js.*

Should also ensure fully qualified names which are covered by an import statement are also flagged.

## Context
Saw `kotlin.io.println()` and thought it was unnecessary and can be simplified to `println()` in https://github.com/arturbosch/detekt/blob/4a08c7c3d5da5f0c09fe030b7628802eee03814a/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/DetektProgressListener.kt

Also see https://pmd.github.io/latest/pmd_rules_java_codestyle.html#unnecessaryfullyqualifiedname

**Comment by BraisGabin** [comment]
I wrote something for this rule, it doesn't work yet. But this rule is a big one so I want to share my thoughts and get some feedback:

- We need to check the package of the file and check for fully qualified name that start with it.
  - and then ensure that there is not an import that force you to use it.
- We need to handle the alias imports `as`
- We need to handle the default imports as it's pointed out by @3flex

Open question:
- Do we want to remove the fully qualified `java.util.HashMap` and similar ones? They are redundant, you can use `HashMap` directly.
- What about `java.util.List`? They are not the same as `List` but you can't be sure that it's a `ModifiableList` neigther. Maybe we need another rule for this.

**Comment by 3flex** [comment]
My original suggested approach was pretty naive, ideally the compiler can provide all the info. Instead I suggest a port of https://github.com/JetBrains/kotlin/blob/6b5d92a6935146391b887319ab752e91febbd4dc/idea/src/org/jetbrains/kotlin/idea/inspections/RemoveRedundantQualifierNameInspection.kt#L35-L142

**Comment by BraisGabin** [comment]
Ok, this rule is bigger than I expected. I have something: https://github.com/BraisGabin/detekt/tree/unnecessary-fully-qualified-name but it's far from done. The part that I like more right now [are the tests](https://github.com/BraisGabin/detekt/blob/unnecessary-fully-qualified-name/detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFullyQualifiedNameSpec.kt). I extracted them from [all this idea test cases](https://github.com/JetBrains/kotlin/tree/6b5d92a6935146391b887319ab752e91febbd4dc/idea/testData/inspectionsLocal/removeRedundantQualifierName).

I probably should port the code from JetBrains but I don't know what to change to get that code working inside detekt. If someone wants to follow this task feel free because I don't think that I would have the time to do it in short term.

**Comment by 3flex** [comment]
Very nice, I generally prefer writing the rule to the tests - thanks for doing that! I'll take your branch and port the rule if that's OK with you?

**Comment by BraisGabin** [comment]
> I just attempted this but unfortunately the IDEA inspection uses a lot of underlying code from the Kotlin plugin for IDEA, which isn't available to the compiler (or detekt).

Yes that was what I found. I'm going to open an issue to talk about wrapping intellij inspections in the same way that we wrap ktlint.

**Comment by cortinico** [comment]
> Would have an easy way to "when lint detect it, please remove"?

There is no easy way to wrap around Lint as we are agnostic to that tool.

It's worth checking if the warning you're seeing is actually from Android Lint, from the compiler of from IntelliJ.

Regardless, writing such a rule shouldn't be that hard 🤔

**Comment by marschwar** [comment]
I think this is a duplicate of #3699 (or the other way around)  and I do not think it is very easy. See [this comment](https://github.com/detekt/detekt/issues/3699#issuecomment-1070978544).

**Comment by 3flex** [comment]
Not a duplicate: this is about unnecessary fully qualified names, while the other issue is about redundant imports.

**Comment by trietbui85** [comment]
Hi, any update on this rule?

**Comment by marschwar** [comment]
> Not a duplicate: this is about unnecessary fully qualified names, while the other issue is about redundant imports.

While it is not a duplicate the comment I referenced also applies here. I do not think it is possible to know for sure that the fully qualified name can be omitted if there is a function by the same name in a different file in the same package.

**Comment by abora97** [comment]
Hi, any update on this rule?

**Comment by BraisGabin** [comment]
No news, PRs are more than welcome. Jetbrains moved this rule to another repo so if someone wants to go for it here is the latest version: https://github.com/JetBrains/intellij-community/blob/master/plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/RemoveRedundantQualifierNameInspection.kt

Also they moved their tests. They are here: https://github.com/JetBrains/intellij-community/tree/23bb68de04cfd849d615dac6ceaa2738e5bd431d/plugins/kotlin/idea/tests/testData/inspectionsLocal/removeRedundantQualifierName

**Comment by BraisGabin** [comment]
> Will this include checking for unnecessary `this` or would `UnneccessaryThis` be its own rule?

If we implement that rule it would be its own rule.

**Comment by forketyfork** [comment]
I've implemented this rule in https://github.com/detekt/detekt/pull/8917, please check it out.
